### PR TITLE
[OpenXLA] Migrate to pull XLA from OpenXLA

### PR DESCRIPTION
**Pull request by ManfeiBai**
**[Description]**
PyTorch/XLA migrate to pull XLA from OpenXLA by replacing TensorFlow with OpenXLA after deprecating XRT usage, and replace TensorFlow-pin with OpenXLA-pin to May09

**[What has changed]**
1. change from `@org_tensorflow//tensorflow/compiler/xla/...` to `@xla//xla/...`
2. change from `@org_tensorflow//tensorflow/tsl/...` to `@tsl//tsl/...`
3. change from `#include "tensorflow/compiler/xla/...` to `#include "xla/...`
4. change from `#include "tensorflow/tsl/...` to `#include "tsl/...`
5. change from pull XLA from [TensorFlow](https://github.com/tensorflow/tensorflow) with `tf_patches` to pull XLA from [OpenXLA](https://github.com/openxla/xla) with `openxla_patches`

**[Start to do OpenXLA-pin]**
After this PR, we will update OpenXLA-pin instead of TensorFlow-pin. We used to update `WORKSPACE` and `tf_patches` to update used TensorFlow's version, aka TensorFlow-pin update. After this PR, we will update OpenXLA-pin by updating `WORKSPACE` and `openxla_patches` to update used OpenXLA's version.

Thanks, @miladm @JackCaoG @will-cromar @puneith @ddunl @stgpetrovic @wonjoolee95 @lsy323 @aspirido
According to https://github.com/pytorch/xla/pull/5202#discussion_r1237741669, we should give AWS folks a heads up when we merge this pr.

### CODE REVIEW

**Comment by ManfeiBai** [comment]
`xla/service/gpu/runtime/executable.cc` skip one change

**Comment by ManfeiBai** [comment]
skip `pprof.BUILD`

**Comment by ManfeiBai** [comment]
found same error like https://github.com/pytorch/xla/pull/5202/commits/567657f4ccd7761ad1aa26ed585964060a76f1a0 in https://github.com/tensorflow/tensorflow/issues/58851#issuecomment-1384305876

**Review by miladm — COMMENTED** [review]
Thanks. I took a quick pass at the PR. Will wait for the CI tests to pass and review.

@ManfeiBai please add a few lines in the PR description explaining the process of updating TorchXLA dependency to openXLA via pin-updates.

**Comment by JackCaoG** [comment]
Hmm TPU CI seems to failed with some linker error
```
Step #2 - "build_pytorch-xla-test_docker_image":     ERROR: /src/pytorch/xla/BUILD:1:10: Linking _XLAC.so failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc @bazel-out/k8-opt/bin/_XLAC.so-2.params
Step #2 - "build_pytorch-xla-test_docker_image":     /usr/bin/ld.gold: error: bazel-out/k8-opt/bin/_solib_k8/_U@local_Uconfig_Ucuda_S_Scuda_Ccudart___Ucuda_Scuda_Slib/libcudart.so: file is empty
Step #2 - "build_pytorch-xla-test_docker_image":     collect2: error: ld returned 1 exit status
```

**Comment by ManfeiBai** [comment]
> Hmm TPU CI seems to failed with some linker error
> 
> ```
> Step #2 - "build_pytorch-xla-test_docker_image":     ERROR: /src/pytorch/xla/BUILD:1:10: Linking _XLAC.so failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc @bazel-out/k8-opt/bin/_XLAC.so-2.params
> Step #2 - "build_pytorch-xla-test_docker_image":     /usr/bin/ld.gold: error: bazel-out/k8-opt/bin/_solib_k8/_U@local_Uconfig_Ucuda_S_Scuda_Ccudart___Ucuda_Scuda_Slib/libcudart.so: file is empty
> Step #2 - "build_pytorch-xla-test_docker_image":     collect2: error: ld returned 1 exit status
> ```

thanks, Jack, yeah, this error appear after adding `"@xla//xla/stream_executor:cuda_platform",` to BUILD's dep, is debuging this error now in https://github.com/pytorch/xla/pull/5209

**Comment by JackCaoG** [comment]
hmm, for a TPUVM environment, we shouldn't need any cuda bits, maybe only include this dependency on cuda build?

**Comment by ManfeiBai** [comment]
> hmm, for a TPUVM environment, we shouldn't need any cuda bits, maybe only include this dependency on cuda build?

Thanks, Jack, tested in https://github.com/pytorch/xla/pull/5209 with only include `"@xla//xla/stream_executor:cuda_platform",` in cuda build:
```
    ] + if_cuda_is_configured([
        "@xla//xla/stream_executor:cuda_platform",
    ]),
```

**Comment by JackCaoG** [comment]
Thanks! ..I am curious why XRT CI also passed, I kind of expected it to fail

**Comment by ManfeiBai** [comment]
> Thanks! ..I am curious why XRT CI also passed, I kind of expected it to fail

iiuc, XRT CI = non-XRT CI + XRT tests,

after we deleted almost all code of XRT, we deleted code added by `disable_xrt`, so we skipped all `disable_xrt` part;

now XRT CI is doing the same thing like non-XRT CI, so it passed too;

because OpenXLA don't support XRT, so there is no XRT code, so every build and tests build and passed without XRT, using OpenXLA

please correct me if I'm wrong, @will-cromar

**Comment by JackCaoG** [comment]
Ok that make sense, this CI is doing more than I thought. My impression is we only disable the XRT but you actually deleted it in this pr. I guess it make sense since you want to delete some TPU topo stuff. I will try to review it tmr, but we can't merge this pr until we decide the future of the XRT. Great work, thanks @ManfeiBai !

File: WORKSPACE
Comment (JackCaoG) [inline WORKSPACE:54]: do we still need all of these patches after we moved to OpenXLA? I thought a lot of them are tf specified

File: configuration.yaml
Comment (JackCaoG) [inline configuration.yaml:98]: I think auto-replace hit some bug here

File: docker/common.sh
Comment (JackCaoG) [inline docker/common.sh:4]: maybe `PJRT_DEVICE=TPU`? the above command is to configure XRT CPU

File: openxla_patches/thread_local_random.diff
Comment (JackCaoG) [inline openxla_patches/thread_local_random.diff:1]: if I remember correctly this is a XRT related patch we can just remove...

File: openxla_patches/grpc_version.diff
Comment (JackCaoG) [inline openxla_patches/grpc_version.diff:1]: this is for XRT only

File: WORKSPACE
Comment (ManfeiBai) [reply WORKSPACE:54]: all the patches are applied to OpenXLA according to current TF-pin;

according to our branch `newbranchopxla`:
- `openxla_patches/optimized_function_graph.diff` is empty now, could delete this file
- `openxla_patches/xplane.diff` is empty now, could delete this file

other patch file would need confirmed before delete

File: configuration.yaml
Comment (ManfeiBai) [reply configuration.yaml:98]: sry, didn't follow, what kind of bug?

File: docker/common.sh
Comment (ManfeiBai) [reply docker/common.sh:4]: Thanks for pointing, test in https://github.com/pytorch/xla/pull/5220

File: openxla_patches/thread_local_random.diff
Comment (ManfeiBai) [reply openxla_patches/thread_local_random.diff:1]: thanks for pointing, test in https://github.com/pytorch/xla/pull/5221

File: openxla_patches/grpc_version.diff
Comment (ManfeiBai) [reply openxla_patches/grpc_version.diff:1]: thanks for pointing, test in https://github.com/pytorch/xla/pull/5222

File: configuration.yaml
Comment (JackCaoG) [reply configuration.yaml:98]: you don't need `RT_WORKERS` and `RT_SHARD_ORDINAL`, they should be removed

Comment (JackCaoG) [reply configuration.yaml:98]: Did you remove the logic to acquire these env var? I think they are just some XRT thing that can be removed.

File: test/pjrt/test_runtime.py
Comment (JackCaoG) [inline test/pjrt/test_runtime.py:69]: you should just error out here, PJRT will become the only runtime, not setting it is an error

File: test/run_tests.sh
Comment (JackCaoG) [inline test/run_tests.sh:136]: why are these removed?

File: test/test_xla_dist.py
Comment (JackCaoG) [inline test/test_xla_dist.py:590]: test_xla_dist can be removed, `xla_dist` only works for xrt

File: torch_xla/core/xla_env_vars.py
Comment (JackCaoG) [inline torch_xla/core/xla_env_vars.py:8]: do we still need these? A lot of them are used by XRT only. can you find out who set these env var?

File: torch_xla/core/xla_model.py
Comment (JackCaoG) [inline torch_xla/core/xla_model.py:143]: I think you can just error out here. @will-cromar to confirm

File: torch_xla/csrc/runtime/env_vars.cc
Comment (JackCaoG) [inline torch_xla/csrc/runtime/env_vars.cc:19]: again I suspect we don't need these XRT_* env var, @will-cromar  to confirm

File: torch_xla/csrc/runtime/metrics_reader.h
Comment (JackCaoG) [inline torch_xla/csrc/runtime/metrics_reader.h:17]: you can just call it `metrics` instead of `rt_metrics`

File: torch_xla/csrc/runtime/topology.proto
Comment (JackCaoG) [inline torch_xla/csrc/runtime/topology.proto:26]: why do we need this? We should be able to remove mesh service after XRT. PJRT does not use mesh service.

File: test/pjrt/test_runtime.py
Comment (ManfeiBai) [reply test/pjrt/test_runtime.py:69]: thanks for pointing, test in https://github.com/pytorch/xla/pull/5223

File: test/run_tests.sh
Comment (ManfeiBai) [reply test/run_tests.sh:136]: `run_async_scalar` is only used in `run_xrt_tests`,
`run_torchrun` is only used in `run_xrt_tests` too :
 https://github.com/pytorch/xla/blob/b3f4c7bc1e9f812069770c83e38e5bd2dc20aef6/test/run_tests.sh#LL138C10-L138C23 

so deleted `run_async_scalar` and `run_torchrun` when delete `run_xrt_tests`

File: torch_xla/distributed/xla_backend.py
Comment (JackCaoG) [inline torch_xla/distributed/xla_backend.py:50]: Note, we should give AWS folks a heads up when we merge this pr.

File: torch_xla/distributed/xla_multiprocessing.py
Comment (JackCaoG) [inline torch_xla/distributed/xla_multiprocessing.py:58]: a lot of functions in this file is only used by XRT, we should do a clean up afterward.

Comment (JackCaoG) [reply torch_xla/distributed/xla_multiprocessing.py:58]: @will-cromar to review change in this file

File: test/test_xla_dist.py
Comment (ManfeiBai) [reply test/test_xla_dist.py:590]: thanks for pointing, test in https://github.com/pytorch/xla/pull/5224

File: torch_xla/core/xla_env_vars.py
Comment (ManfeiBai) [reply torch_xla/core/xla_env_vars.py:8]: these var are added 4 years ago:
https://github.com/pytorch/xla/blame/b3f4c7bc1e9f812069770c83e38e5bd2dc20aef6/torch_xla/core/xla_env_vars.py#LL7-L9C30

could create a ticket to track whether we still need these env var or not

File: torch_xla/core/xla_model.py
Comment (ManfeiBai) [reply torch_xla/core/xla_model.py:143]: thanks for pointing, test in https://github.com/pytorch/xla/pull/5228

File: torch_xla/csrc/runtime/env_vars.cc
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/env_vars.cc:19]: thanks, will track in ticket too

File: torch_xla/csrc/runtime/metrics_reader.h
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/metrics_reader.h:17]: thanks for pointing, test in https://github.com/pytorch/xla/pull/5229

File: torch_xla/csrc/runtime/topology.proto
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/topology.proto:26]: PJRT has a test for mesh service: https://github.com/pytorch/xla/blob/master/test/pjrt/test_mesh_service.py

not sure whether this test is still needed or not

File: torch_xla/distributed/xla_backend.py
Comment (ManfeiBai) [reply torch_xla/distributed/xla_backend.py:50]: got it, have added in the list of full list of OpenXLA migration

File: torch_xla/distributed/xla_multiprocessing.py
Comment (ManfeiBai) [reply torch_xla/distributed/xla_multiprocessing.py:58]: sure, will do and saved in the list of full list of OpenXLA miration list

File: WORKSPACE
Comment (ManfeiBai) [reply WORKSPACE:54]: since https://github.com/pytorch/xla/pull/5219 has passed all tests, will apply this change later

File: docker/common.sh
Comment (ManfeiBai) [reply docker/common.sh:4]: #5220's change has been saved in https://github.com/pytorch/xla/pull/5222 too, since https://github.com/pytorch/xla/pull/5222 has passed all tests, will apply this change in #5220

File: openxla_patches/thread_local_random.diff
Comment (ManfeiBai) [reply openxla_patches/thread_local_random.diff:1]: #5221's change has been saved in https://github.com/pytorch/xla/pull/5222 too, since https://github.com/pytorch/xla/pull/5222 has passed all tests, will apply this change in #5221

File: openxla_patches/grpc_version.diff
Comment (ManfeiBai) [reply openxla_patches/grpc_version.diff:1]: https://github.com/pytorch/xla/pull/5222 has passed all tests, will apply this change

File: test/pjrt/test_runtime.py
Comment (ManfeiBai) [reply test/pjrt/test_runtime.py:69]: try to raise an error here, but failed in https://github.com/pytorch/xla/pull/5223/commits/cd0f293c586e55e75cd3b573241d3ca9533cd2b0

since this is a test to test pjrt,  we have several situation here for PJRT test:
-   ('default', /*env_vars=*/{}, /*expect_using_pjrt=*/True), 
  - this set is testing whether we have PJRT as default
-   ('no_default', /*env_vars=*/{'PJRT_SELECT_DEFAULT_DEVICE': '0'}, /*expect_using_pjrt=*/False), 
  - this set is no PJRT_DEVICE set, and no device get
-  ('pjrt_cpu', /*env_vars=*/{'PJRT_DEVICE': 'CPU', 'PJRT_SELECT_DEFAULT_DEVICE': '0'}, /*expect_using_pjrt=*/True), 
  - PJRT_DEVICE=CPU test, will keep
-  ('xrt_tpu', /*env_vars=*/{'XRT_TPU_CONFIG': 'localservice;0;localhost:51011'}, /*expect_using_pjrt=*/False), 
  - this set has been deleted, no XRT code and no XRT TPU test
-  ('pjrt_tpu_precedence', /*env_vars=*/{'PJRT_DEVICE': 'TPU', 'XRT_TPU_CONFIG': 'localservice;0;localhost:51011',}, /*expect_using_pjrt=*/True), 
  - PJRT_DEVICE=TPU test, will keep
-  ('xrt_gpu', /*env_vars=*/{'GPU_NUM_DEVICES': '4'}, /*expect_using_pjrt=*/False), 
  - this set has been delelted, no XRT code and no XRT GPU test
-  ('pjrt_gpu', /*env_vars=*/{'PJRT_DEVICE': 'GPU','GPU_NUM_DEVICES': '4'}, /*expect_using_pjrt=*/True), 
  - PJRT_DEVICE=GPU test, will keep
-  ('xla_dist_worker', /*env_vars=*/{'XRT_LOCAL_WORKER': 'c_localservice:2'}, /*expect_using_pjrt=*/False))
  - this set will be delete, `xla_dist` is for XRT, and will delete XRT's xla_dist test


for situation which PJRT is not set and still want PJRT, we will set default PJRT_DEVICE to `CPU` or `TPU` based on whether or not found libtpu.so and TPU device: https://github.com/pytorch/xla/blob/b3f4c7bc1e9f812069770c83e38e5bd2dc20aef6/torch_xla/runtime.py#L36C1-L51C41

so we should not raise ERROR here for each situation

File: test/test_xla_dist.py
Comment (ManfeiBai) [reply test/test_xla_dist.py:590]: since #5224 has passed all tests, will apply this change

**Comment by will-cromar** [comment]
Thanks @ManfeiBai. I'm very excited to see your progress here!

There are three separate changes that I would like to be submitted separately:

1. Delete XRT from the main branch
2. Swap TensorFlow and OpenXLA
3. Update pin to the latest XLA and libtpu

(2) and (3) can happen in any order after (1)

All three of these changes have significant risks. I expect each one to require a rollback for unforeseen issues, and I'd like to be able to roll back each step incrementally if needed.

I'll prepare a separate PR for (1) that deletes XRT since I have a good idea of where the dead code will be.

**Comment by ManfeiBai** [comment]
> Thanks @ManfeiBai. I'm very excited to see your progress here!
> 
> There are three separate changes that I would like to be submitted separately:
> 
> 1. Delete XRT from the main branch
> 2. Swap TensorFlow and OpenXLA
> 3. Update pin to the latest XLA and libtpu
> 
> (2) and (3) can happen in any order after (1)
> 
> All three of these changes have significant risks. I expect each one to require a rollback for unforeseen issues, and I'd like to be able to roll back each step incrementally if needed.
> 
> I'll prepare a separate PR for (1) that deletes XRT since I have a good idea of where the dead code will be.

Thanks, @will-cromar, will rebase with your (1) PR after it merged

File: docker/common.sh
Comment (ManfeiBai) [reply docker/common.sh]: Thanks, updated

File: openxla_patches/thread_local_random.diff
Comment (ManfeiBai) [reply openxla_patches/thread_local_random.diff]: Thanks, updated

File: openxla_patches/grpc_version.diff
Comment (ManfeiBai) [reply openxla_patches/grpc_version.diff]: Thanks, updated

File: test/pjrt/test_runtime.py
Comment (ManfeiBai) [reply test/pjrt/test_runtime.py]: Thanks, updated by deleting this print comments

File: test/test_xla_dist.py
Comment (ManfeiBai) [reply test/test_xla_dist.py:590]: Thanks, updated

File: torch_xla/core/xla_model.py
Comment (ManfeiBai) [reply torch_xla/core/xla_model.py]: Thanks, updated

File: torch_xla/csrc/runtime/env_vars.cc
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/env_vars.cc:19]: Thanks, updated

File: torch_xla/csrc/runtime/metrics_reader.h
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/metrics_reader.h]: Thanks, updated

File: torch_xla/csrc/runtime/topology.proto
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/topology.proto:26]: mesh service would been deleted in https://github.com/pytorch/xla/pull/5240#pullrequestreview-1495708204, will delete `topology.proto` after rebase with merged #5240

**Comment by ManfeiBai** [comment]
fixed according to comments in this PR and https://github.com/pytorch/xla/pull/5222

will rebase with https://github.com/pytorch/xla/pull/5240/files later

File: bazel/openxla.bzl
Comment (ManfeiBai) [inline bazel/openxla.bzl:1]: delete this file after rebased with#5240

File: configuration.yaml
Comment (ManfeiBai) [reply configuration.yaml:98]: thanks for clarification, will delete these three envs `RT_SHARD_LOCAL_ORDINAL`, `RT_SHARD_ORDINAL` and `RT_WORKERS` after rebase with #5240

**Comment by ManfeiBai** [comment]
check `rt_metrics` and `GetMetricRenderInfo()` in `[torch_xla/csrc/runtime/metrics_reader.cc](https://github.com/pytorch/xla/pull/5222/files/5b94b8b4ec33bab9eac08a90cb26af659ffcc3e4#diff-5c3a4f94c1a72a1ad2de9ad677e8f288ccaf1afe01108eddb771daa927b571f8)` according to @miladm's comment in https://github.com/pytorch/xla/pull/5222#discussion_r1239121138

File: bazel/openxla.bzl
Comment (ManfeiBai) [reply bazel/openxla.bzl:1]: rebased and keep it for a short time, would delete in the future

**Comment by ManfeiBai** [comment]
`CreateXrtMetricReport` is deleted and `GetMetricRenderInfo` is kept for now, leave this as open discuss for now, and get back to it after rebased with #5240

**Comment by ManfeiBai** [comment]
rebased #5240 with and testing in https://github.com/pytorch/xla/pull/5242, once all tests passed in #5242, will update @5202 too

File: configuration.yaml
Comment (will-cromar) [reply configuration.yaml:42]: It looks like these got added back in the merge. Please remove the `RT` config vars.

File: docs/bazel.md
Comment (ManfeiBai) [reply docs/bazel.md:13]: we have deleted these two patches in this PR, so keep these two patches in this md might confuse;

but this .md file is not updated at all, so we should update this document in a follow-up, and that would be clear too

File: torch_xla/csrc/runtime/sys_util.h
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/sys_util.h:20]: thanks, updated

File: torch_xla/csrc/runtime/metrics_analysis.cc
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/metrics_analysis.cc:93]: Thanks, updated and testing

File: test/pjrt/test_runtime.py
Comment (ManfeiBai) [reply test/pjrt/test_runtime.py:11]: thanks, updated

File: docker/common.sh
Comment (ManfeiBai) [reply docker/common.sh:1]: thanks, updated

File: configuration.yaml
Comment (ManfeiBai) [reply configuration.yaml:42]: thanks, updated

File: torch_xla/distributed/xla_dist.py
Comment (ManfeiBai) [reply torch_xla/distributed/xla_dist.py:1]: thanks, updated and testing in https://github.com/pytorch/xla/pull/5260

File: torch_xla/distributed/xrt_init.py
Comment (ManfeiBai) [reply torch_xla/distributed/xrt_init.py:1]: thanks, updated and testing in https://github.com/pytorch/xla/pull/5260

File: torch_xla/core/xla_env_vars.py
Comment (ManfeiBai) [reply torch_xla/core/xla_env_vars.py]: has rebased and these veriables has been deleted before rebase

File: torch_xla/csrc/runtime/topology.proto
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/topology.proto]: has rebased to Jun27, topology has been deleted before rebase

File: torch_xla/distributed/xla_multiprocessing.py
Comment (ManfeiBai) [reply torch_xla/distributed/xla_multiprocessing.py]: has revert this change and @will-cromar will have follow-up PR for XRT contain this change

File: bazel/openxla.bzl
Comment (ManfeiBai) [reply bazel/openxla.bzl:1]: testing in https://github.com/pytorch/xla/pull/5260

File: test/pjrt/test_runtime.py
Comment (ManfeiBai) [reply test/pjrt/test_runtime.py]: thanks, updated

**Review by will-cromar — APPROVED** [review]
LGTM

**Comment by ddunl** [comment]
Nice work!

File: torch_xla/csrc/runtime/openxla_exported_symbols.lds
Comment (JackCaoG) [inline torch_xla/csrc/runtime/openxla_exported_symbols.lds:1]: I have a feeling that a lot of these won't be needed anymore, but we can clean it up later

Comment (ManfeiBai) [reply torch_xla/csrc/runtime/openxla_exported_symbols.lds:1]: sure, thanks for pointing, will add in the list too, will have follow-up PR to update doc and test to delete these lds files too

**Review by JackCaoG — APPROVED** [review]
Thanks @ManfeiBai ! This looks great! Let's try to merge this pr tmr!

File: torch_xla/csrc/runtime/openxla_exported_symbols.lds
Comment (ManfeiBai) [reply torch_xla/csrc/runtime/openxla_exported_symbols.lds:1]: This file has been deleted in https://github.com/pytorch/xla/pull/5275

**Pull request merged**
